src/libstd/io/mod.rs

```diff
@@ -1549,93 +1594,182 @@ impl<R: Read> Iterator for Bytes<R> {
 
 /// An iterator over the `char`s of a reader.
 ///
-/// This struct is generally created by calling [`chars()`][chars] on a reader.
-/// Please see the documentation of `chars()` for more details.
+/// This struct is generally created by calling [`utf8_chars()`][utf8_chars] on a reader.
+/// Please see the documentation of `utf8_chars()` for more details.
 ///
-/// [chars]: trait.Read.html#method.chars
-#[unstable(feature = "io", reason = "awaiting stability of Read::chars",
+/// [utf8_chars]: trait.BufRead.html#method.utf8_chars
+#[unstable(feature = "io", reason = "awaiting stability of BufRead::utf8_chars",
            issue = "27802")]
-pub struct Chars<R> {
+pub struct Utf8Chars<R> {
     inner: R,
 }
 
-/// An enumeration of possible errors that can be generated from the `Chars`
+/// An enumeration of possible errors that can be generated from the `Utf8Chars`
 /// adapter.
 #[derive(Debug)]
-#[unstable(feature = "io", reason = "awaiting stability of Read::chars",
+#[unstable(feature = "io", reason = "awaiting stability of BufRead::utf8_chars",
            issue = "27802")]
-pub enum CharsError {
+pub enum Utf8CharsError {
     /// Variant representing that the underlying stream was read successfully
-    /// but it did not contain valid utf8 data.
-    NotUtf8,
+    /// but contains a byte sequence ill-formed in UTF-8.
+    InvalidUtf8,
+
+    /// Variant representing that the underlying stream contains the start
+    /// of a byte sequence well-formed in UTF-8, but ends prematurely.
+    IncompleteUtf8,
 
     /// Variant representing that an I/O error occurred.
-    Other(Error),
+    Io(Error),
 }
 
-#[unstable(feature = "io", reason = "awaiting stability of Read::chars",
+#[unstable(feature = "io", reason = "awaiting stability of BufRead::utf8_chars",
            issue = "27802")]
-impl<R: Read> Iterator for Chars<R> {
-    type Item = result::Result<char, CharsError>;
+impl<R: BufRead> Iterator for Utf8Chars<R> {
+    type Item = result::Result<char, Utf8CharsError>;
+
+    fn next(&mut self) -> Option<result::Result<char, Utf8CharsError>> {
+        macro_rules! read_byte {
+            (EOF => $on_eof: expr) => {
+                {
+                    let byte;
+                    loop {
+                        match self.inner.fill_buf() {
+                            Ok(buffer) => {
+                                if let Some(&b) = buffer.first() {
+                                    byte = b;
+                                    break
+                                } else {
+                                    $on_eof
+                                }
+                            }
+                            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
```

Comment:
Why should it not be useful?
Without it there's no possibility to interrupt a blocking call.

I already had to reimplement a large part of `std::io` in my crate [netio](http://troplin.bitbucket.org/netio-rs/0.3.1/netio/index.html), just to be able to interrupt those calls (and not lose any data).
If you remove it also from `Read::read`, `BufRead::fill_buf` and `Write::write`, then this will be impossible without reimplementing everything from scratch!